design/metadata-handling.md

```diff
@@ -0,0 +1,286 @@
+`worker-np1-0`, `worker-np1-1`, `worker-np1-2` etc. and those names to be
+unrelated with the BareMetalHost names (that might be `machine-xyz`, reflecting
+some hardware related information, not Kubernetes node information).
+Hence we want a way to generate a node-specific metadata for a machine
+deployment and pass it down to Ironic as metadata.
+
+### Goals
+
+The goals are :
+
+- to add a metadata field in BareMetalHost to have it configurable
+  for Ironic, and in the Metal3Machine for CAPM3 to be able to pass it down to
+  BareMetalHost.
+- introduce new objects containing information to get the metadata per node,
+  or generate it, and store the values in use to not duplicate them.
+- The reconciliation of the Metal3Machine would use those metadata objects to
+  properly fill the metadata field of BareMetalHost
+
+
+### Non-Goals
+
+TBA
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to define a set of node-specific element to give as metadata
+through ironic to fill the cloud-init template. Those elements should be used
+for one node only, and be re-used during upgrade.
+
+#### Story 2
+
+As a user, I want to give my own metadata to Ironic as a field on BareMetalHost.
+
+### Implementation Details/Notes/Constraints
+
+- Baremetal Operator should set some default metadata in case they are not
+  provided by the use (Like UUID that is required).
+- Not providing any metadata object and any content in the metadata fields of
+  Metal3Machine should result in the same behaviour as before this proposal is
+  implemented
+- the metadata template field would be a list of objects giving the key of the
+  metadata and information on how to retrieve the value, such as an object
+  reference to the metadata object and the key in that object to use.
+- Documentation of this feature would be very important.
+- The controllers must be able to recreate all status objects, i.e. no
+  necessary information should be stored in the status without being recoverable
+
+### Risks and Mitigations
+
+From a user point of view, if this is not well documented or understood, it
+could lead to unexpected errors or behaviours. The errors should be easily
+understandable.
+
+## Design Details
+
+BareMetalHost would have an added field `metaData` in the `Spec` pointing to a
+secret containing the metadata map. Bare Metal Operator would then pass this map
+with added default key/value pairs if not present to Ironic.
+
+A new object would be created, a Metal3Metadata type.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Metadata
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  values:
+    node-0:
+      abc: def
+    node-1:
+      abc: ghi
+  templates:
+    local-hostname: worker-np1-{{ getIndex }}
+```
```

Comment:
This looks like a ConfigMap. Why do we need a new CRD?

It looks like data for multiple hosts is mixed together in one instance here. Would we really do that? Why not use a separate resource for each host?

Reply:
The idea is that this resource would be tied to a machine deployment, or a kubeadm control plane object. That's why there are multiple hosts, since they are automatically created for the machine deployment or KCP.

Yes in the first iteration, this was a configmap, but we benefit from using CR, with the validation opportunity that it brings, not only with openapi, but also with the webhooks.

Reply:
In the new proposal, I have a streamlined the object, using only a template, to render the metadata for all machines referencing this object.